plugin/plugins/filesystem/filesystem.go

```diff
@@ -0,0 +1,124 @@
+package filesystem
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+
+	"github.com/djdv/p9/p9"
+	plugin "github.com/ipfs/go-ipfs/plugin"
+	fsnodes "github.com/ipfs/go-ipfs/plugin/plugins/filesystem/nodes"
+	logging "github.com/ipfs/go-log"
+	coreiface "github.com/ipfs/interface-go-ipfs-core"
+	"github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr-net"
+)
+
+var (
+	_ plugin.PluginDaemon = (*FileSystemPlugin)(nil) // impl check
+
+	// Plugins is an exported list of plugins that will be loaded by go-ipfs.
+	Plugins = []plugin.Plugin{
+		&FileSystemPlugin{}, //TODO: individually name implementations: &P9{}
+	}
+
+	logger logging.EventLogger
+)
+
+func init() {
+	logger = logging.Logger("plugin/filesystem")
+}
+
+type FileSystemPlugin struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	addr      multiaddr.Multiaddr
+	listener  manet.Listener
+	errorChan chan error
+}
+
+func (*FileSystemPlugin) Name() string {
+	return PluginName
+}
+
+func (*FileSystemPlugin) Version() string {
+	return PluginVersion
+}
+
+func (fs *FileSystemPlugin) Init(env *plugin.Environment) error {
+	logger.Info("Initializing 9P resource server...")
+	if !filepath.IsAbs(env.Repo) {
+		absRepo, err := filepath.Abs(env.Repo)
+		if err != nil {
+			return err
+		}
+		env.Repo = absRepo
+	}
+
+	cfg := &Config{}
+	// config not being set is okay and will load defalts
+	// config being set with malformed data is not okay and will instruct the daemon to halt-and-catch-fire
+	rawConf, ok := (env.Config).(json.RawMessage)
+	if ok {
+		if err := json.Unmarshal(rawConf, cfg); err != nil {
+			return err
+		}
+	} else {
+		if env.Config != nil {
+			return fmt.Errorf("plugin config does not appear to be correctly formatted: %#v", env.Config)
+		}
+		cfg = defaultConfig(env.Repo)
+	}
+
+	var err error
+	if envAddr := os.ExpandEnv(EnvAddr); envAddr == "" {
+		fs.addr, err = multiaddr.NewMultiaddr(cfg.Service[defaultService])
+	} else {
+		fs.addr, err = multiaddr.NewMultiaddr(envAddr)
+	}
+	if err != nil {
+		return err
+	}
+
+	logger.Info("9P resource server okay for launch")
+	return nil
+}
+
+func (fs *FileSystemPlugin) Start(core coreiface.CoreAPI) error {
+	logger.Info("Starting 9P resource server...")
+
+	//TODO [manet]: unix sockets are not removed on process death (on Windows)
+	// so for now we just try to remove it before listening on it
+	if runtime.GOOS == "windows" {
+		removeUnixSockets(fs.addr)
+	}
+
+	fs.ctx, fs.cancel = context.WithCancel(context.Background())
+	fs.errorChan = make(chan error, 1)
+
+	var err error
+	if fs.listener, err = manet.Listen(fs.addr); err != nil {
```

Comment:
Avoid "if assignment" here (it has side effects).

Reply:
Res: https://github.com/ipfs/go-ipfs/pull/6612/commits/eaff0c93b4bf15113cdf02a519c584ae77aae89e#diff-8d243ab8514fa1b77d41df37a81641c9R103